actionview/lib/action_view/unbound_template.rb

```diff
@@ -14,10 +14,26 @@ def initialize(source, identifier, details:, virtual_path:)
       @virtual_path = virtual_path
 
       @templates = Concurrent::Map.new(initial_capacity: 2)
+      @write_lock = Mutex.new
+      @templates = {}
     end
 
     def bind_locals(locals)
-      @templates[locals] ||= build_template(locals)
+      if template = @templates[locals]
```

Comment:
Should not the read also hold the lock? Otherwise we could be reading from a hash that is being written.

Reply:
It might be subtle, but we do have a double-check in the synchronize block using `||=`, which ensures we won't accidentally duplicate. This uses (now that I fixed a mistake 😳) `Concurrent::Map#[]`, which is unsynchronized on MRI but should make sure the reads and writes themselves are safe.